File: ros2_packages/hello_action_test_cpp/src/hello_client.cpp
// Client side (using Action)
#include <memory>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_action/rclcpp_action.hpp"
#include "hello_msgs/action/hello.hpp"

class HelloClient : public rclcpp::Node
{
public: // alias, constructors, and destructors
    using Hello = hello_msgs::action::Hello;
    using GoalHandleHello = rclcpp_action::ClientGoalHandle<Hello>;

    explicit HelloClient(const rclcpp::NodeOptions &options);
public: // functions
    bool is_goal_done() const;
    void send_goal();
private: // variables and instances
    rclcpp_action::Client<Hello>::SharedPtr action_client;
    rclcpp::TimerBase::SharedPtr timer;
    bool goal_done;
private: // functions
    void goal_responseCB(std::shared_future<GoalHandleHello::SharedPtr> future);
    void feedbackCB(
        GoalHandleHello::SharedPtr,
        const std::shared_ptr<const Hello::Feedback> feedback
    );
    void resultCB(const GoalHandleHello::WrappedResult &result);
};
HelloClient::HelloClient(const rclcpp::NodeOptions &options=rclcpp::NodeOptions())
    : Node("hello_action_test_client", options), goal_done(false)
{
    this->action_client = rclcpp_action::create_client<Hello>(
        this->get_node_base_interface(),
        this->get_node_graph_interface(),
        this->get_node_logging_interface(),
        this->get_node_waitables_interface(),
        "hello_action"
    );
    this->timer = this->create_wall_timer(
        std::chrono::milliseconds(500),
        std::bind(&HelloClient::send_goal, this)
    );
}
bool HelloClient::is_goal_done() const
{
    return this->goal_done;
}
void HelloClient::send_goal()
{
    using namespace std::placeholders;

    this->timer->cancel();
    this->goal_done = false;

    if (!this->action_client)
        RCLCPP_ERROR(this->get_logger(), "Action client not initialized");

    if (!this->action_client->wait_for_action_server(std::chrono::seconds(10)))
    {
        RCLCPP_ERROR(this->get_logger(), "Action server not available after waiting");
        this->goal_done = true;
        return;
    }

    auto goal_msg = Hello::Goal();
    goal_msg.hello_goal = "Hello World!";

    RCLCPP_INFO(this->get_logger(), "Sending goal to server");
    
    auto send_goal_options = rclcpp_action::Client<Hello>::SendGoalOptions();
    send_goal_options.goal_response_callback =
        std::bind(&HelloClient::goal_responseCB, this, _1);
    send_goal_options.feedback_callback =
        std::bind(&HelloClient::feedbackCB, this, _1, _2);
    send_goal_options.result_callback =
        std::bind(&HelloClient::resultCB, this, _1);
    auto goal_handle_future =
        this->action_client->async_send_goal(goal_msg, send_goal_options);
}
void HelloClient::goal_responseCB(std::shared_future<GoalHandleHello::SharedPtr> future)
{
    auto goal_handle = future.get();
    if (!goal_handle)
    {
        RCLCPP_ERROR(this->get_logger(), "Goal rejected from server");
    }
    else
    {
        RCLCPP_INFO(this->get_logger(), "Goal accepted by server, waiting for results");
    }
}
void HelloClient::feedbackCB(
    GoalHandleHello::SharedPtr,
    const std::shared_ptr<const Hello::Feedback> feedback
)
{
    RCLCPP_INFO(this->get_logger(), "Current status: %s", feedback->status.c_str());
}
void HelloClient::resultCB(const GoalHandleHello::WrappedResult &result)
{
    this->goal_done = true;
    switch (result.code)
    {
    case rclcpp_action::ResultCode::SUCCEEDED:
        break;
    case rclcpp_action::ResultCode::ABORTED:
        RCLCPP_ERROR(this->get_logger(), "GOAL ABORTED!");
        return;
    case rclcpp_action::ResultCode::CANCELED:
        RCLCPP_ERROR(this->get_logger(), "GOAL CANCELED!");
        return;
    default:
        RCLCPP_ERROR(this->get_logger(), "UNKNOWN RESULT CODE!");
        break;
    }

    RCLCPP_INFO(this->get_logger(), "Result received\n\tresult: %s", result.result->status.c_str());
}

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);

    auto hello_action_client = std::make_shared<HelloClient>();

    while (!hello_action_client->is_goal_done())
    {
        rclcpp::spin_some(hello_action_client);
    }
    
    rclcpp::shutdown();
}

File: ros2_packages/hello_action_test_py3/hello_action_test_py3/hello_client.py
# hello_client
from action_msgs.msg import GoalStatus
from hello_msgs.action import Hello

import rclpy
from rclpy.action import ActionClient
from rclpy.node import Node

# The Client tells the Server what the Client wants while the Server handles the process
class HelloClient(Node):
    def __init__(self):
        super().__init__('hello_action_test_client')

        self.hello_client = ActionClient(self, Hello, 'hello_action')

    def goal_responseCB(self, future): # dunno what the future handles...
        goal_handle = future.result()
        if not goal_handle.accepted:
            self.get_logger().info('Goal rejected: {0.accepted}'.format(goal_handle))
            return

        self.get_logger().info('Goal accepted: {0.accepted}'.format(goal_handle))

        self.get_result_future = goal_handle.get_result_async()
        self.get_result_future.add_done_callback(self.get_resultCB)

    def feedbackCB(self, feedback):
        # Gets feedback data by "goal_handle.publish_feedback" function from server
        self.get_logger().info('Received feedback:\n{0.status}'.format(feedback.feedback))

    def get_resultCB(self, future):
        result = future.result().result
        status = future.result().status # probably "goal_handle.succeed()" from server?

        if status == GoalStatus.STATUS_SUCCEEDED:
            self.get_logger().info('Goal succeeded!\n\t Result: {0.status}'.format(result))
        else:
            self.get_logger().info('Goal failed with status: {}'.format(status))

    def send_goal(self):
        self.get_logger().info('Waiting for action server...')
        self.hello_client.wait_for_server()

        goal_msg = Hello.Goal()
        goal_msg.hello_goal = 'Hello World!'

        self.get_logger().info('Sending goal request...')
        self.send_goal_future = self.hello_client.send_goal_async(goal_msg,
                feedback_callback=self.feedbackCB)
        self.send_goal_future.add_done_callback(self.goal_responseCB)

    def destroy(self):
        self.hello_client.destroy()
        self.destroy_node()

def main(args=None):
    rclpy.init(args=args)

    hello_client = HelloClient()
    hello_client.send_goal()
    try:
        rclpy.spin(hello_client)
    except KeyboardInterrupt:
        pass
    hello_client.destroy()
    rclpy.shutdown()
